rust/crd/src/lib.rs

```diff
@@ -256,35 +262,32 @@ impl HdfsCluster {
         &self,
         role: &HdfsRole,
         rolegroup_ref: &RoleGroupRef<HdfsCluster>,
-    ) -> (Vec<PersistentVolumeClaim>, ResourceRequirements) {
-        let mut rg_resources = self.rolegroup_resources(role, rolegroup_ref);
+    ) -> Result<(Vec<PersistentVolumeClaim>, ResourceRequirements), Error> {
+        // Initialize the result with all default values as baseline
+        let default_resources = self.default_resources();
+
         let mut role_resources = self.role_resources(role);
-        let mut default_resources = Some(self.default_resources());
-
-        let r = [
-            default_resources.as_mut(),
-            role_resources.as_mut(),
-            rg_resources.as_mut(),
-        ]
-        .into_iter()
-        .flatten()
-        .reduce(|old, new| chainable_merge(new, old))
-        .unwrap(); // always succeeds because default_resources is Some()
-
-        let result_1 = r
+        let mut rg_resources = self.rolegroup_resources(role, rolegroup_ref);
+
+        role_resources.merge(&default_resources);
+        rg_resources.merge(&role_resources);
+
+        let resources: Resources<Storage, NoRuntimeLimits> =
+            fragment::validate(rg_resources).expect("Validation error");
```

Comment:
This is rather internal i agree, but i works with user defined input. I would not let the operator crash here since it may be responsible for other "correct" clusters as well.
```suggestion
        let resources: Resources<Storage, NoRuntimeLimits> = fragment::validate(rg_resources)
            .map_err(|source| Error::FragmentValidationFailure { source })?;
```
And the error in error.rs:
```
    #[error("fragment validation failure")]
    FragmentValidationFailure { source: ValidationError },
```

Reply:
Done - thanks!